src/dgm_solo12.cpp

```diff
@@ -50,37 +50,98 @@ void DGMSolo12::initialize_hardware_communication_process()
         params_["hardware_communication"], "serial_port", serial_port);
 
     solo_.initialize(network_id, serial_port);
-}
+  }
+
+  bool DGMSolo12::is_in_safety_mode()
+  {
+    // Check if any card is in an error state.
+    if (solo_.has_error()) {
+      was_in_safety_mode_ = true;
+      static int counter = 0;
+      if (counter % 2000 == 0) {
+        printf("DGMSolo12: Going into safe mode as motor card reports error.\n");
+      }
+      counter += 1;
+    }
 
-bool DGMSolo12::is_in_safety_mode()
-{
-    static int counter = 0;
-    if (solo_.get_joint_velocities().cwiseAbs().maxCoeff() > 100000003.875)
+    // Check for too fast velocity. Value estiamted from small jump.
+    if (solo_.get_joint_velocities().cwiseAbs().maxCoeff() > 80.)
     {
-        printf(
-            "DGMSolo12: Going into safe mode as joint velocity exceeded "
-            "bound.\n");
-        was_in_safety_mode_ = true;
+      was_in_safety_mode_ = true;
+      static int counter = 0;
+      if (counter % 2000 == 0)
+      {
+        printf("DGMSolo12: Going into safe mode as joint velocity exceeded bound.\n");
+      }
+      counter += 1;
     }
 
-    if (was_in_safety_mode_ || DynamicGraphManager::is_in_safety_mode())
-    {
+    // Check for joint limits. Only do this when calibration is not running.
+    if (!solo_.is_calibrating()) {
+      Eigen::Array<double, 12, 1> lower_lim, upper_lim;
+      lower_lim << -1.2, -1.7, -3.4, -1.2, -1.7, -3.4,
```

Comment:
why not from the xacro/urdf?

Reply:
you want to parse the urdf file and extract the joint limits for each joint? Well, this turned out to be easier/faster to implement.

Reply:
pinocchio do it for you... But ok, why not...